cvmfs/receiver/session_token.cc

```diff
@@ -0,0 +1,179 @@
+
+  const uint64_t current_time = platform_monotonic_time();
+  if (std::numeric_limits<uint64_t>::max() - max_lease_time < current_time) {
+    return 5;
+  }
+
+  const std::string expiry(StringifyUint(current_time + max_lease_time));
+
+  std::string encrypted_body;
+  if (!cipher->Encrypt(
+          "{\"path\" : \"" + path + "\", \"expiry\" : \"" + expiry + "\"}",
+          *secret, &encrypted_body)) {
+    return 6;
+  }
+
+  *session_token = Base64("{\"token_id\" : \"" + *public_token_id +
+                          "\", \"blob\" : \"" + Base64(encrypted_body) + "\"}");
+
+  return 0;
+}
+
+/**
+ * Obtain the public_id from a session token
+ */
+int GetTokenPublicId(const std::string& token, std::string* public_id) {
+  if (public_id == NULL) {
+    return 1;
+  }
+
+  std::string debased64_token;
+  if (!Debase64(token, &debased64_token)) {
+    return 2;
+  }
+
+  UniquePtr<JsonDocument> token_json(JsonDocument::Create(debased64_token));
+  if (!token_json.IsValid()) {
+    return 3;
+  }
+
+  const JSON* token_id =
+      JsonDocument::SearchInObject(token_json->root(), "token_id", JSON_STRING);
+  const JSON* blob =
+      JsonDocument::SearchInObject(token_json->root(), "blob", JSON_STRING);
+
+  if (token_id == NULL || blob == NULL) {
+    return 4;
+  }
+
+  *public_id = token_id->string_value;
+
+  return 0;
+}
+
+/*
+ * Check the validity of a session token using the associated secret
+ */
+int CheckToken(const std::string& token, const std::string& secret,
+               std::string* lease_path) {
+  if (!lease_path) {
+    return 1;
+  }
+
+  std::string debased64_token;
+  if (!Debase64(token, &debased64_token)) {
+    return 2;
+  }
+
+  UniquePtr<JsonDocument> token_json(JsonDocument::Create(debased64_token));
+  if (!token_json.IsValid()) {
+    return 2;
+  }
+
+  const JSON* token_id =
+      JsonDocument::SearchInObject(token_json->root(), "token_id", JSON_STRING);
+  const JSON* blob =
+      JsonDocument::SearchInObject(token_json->root(), "blob", JSON_STRING);
+  if (token_id == NULL || blob == NULL) {
+    return 3;
+  }
+
+  std::string debased64_secret;
+  if (!Debase64(secret, &debased64_secret)) {
+    return 4;
+  }
+  UniquePtr<cipher::Key> key(cipher::Key::CreateFromString(debased64_secret));
+  if (!key.IsValid()) {
+    return 5;
+  }
+
+  std::string encrypted_body;
+  if (!Debase64(blob->string_value, &encrypted_body)) {
+    return 6;
+  }
+
+  std::string body;
+  if (!cipher::Cipher::Decrypt(encrypted_body, *key, &body)) {
+    return 7;
+  }
+
+  UniquePtr<JsonDocument> body_json(JsonDocument::Create(body));
+  if (!token_json.IsValid()) {
+    return 8;
+  }
+
+  const JSON* path =
+      JsonDocument::SearchInObject(body_json->root(), "path", JSON_STRING);
+  const JSON* expiry =
+      JsonDocument::SearchInObject(body_json->root(), "expiry", JSON_STRING);
+  if (path == NULL || expiry == NULL) {
+    return 9;
+  }
+
+  // TODO(radu): can we still use monotonic time if the process restarts?
```

Comment:
Hm, actually we seem to be ok.  On both, Linux and Mac the calls behind `platform_monotonic_time` ensure a stable starting point.